[Bug]: Verify Storage State handling

Hey team,

I'm experimenting with playwright, and trying to save the authenticated state using the `context.StorageState()`
However, struggling to find the way to restore the stored state.

While investigating, found the reference in this [doc](https://playwright.dev/docs/auth/), and API doc for [NewContext](https://playwright.dev/docs/api/class-browser/#browsernewcontextoptions) says that there should be an Option that allows to restore the state, namely `storageState`.

However, browsing the sources it seems that it is not present in the [current implementation](https://github.com/mxschmitt/playwright-go/blob/067f4185d7b5d1f2fffacbc7ac2cfae47b8afe9a/generated-structs.go#L3).

My go.mod
```
// ...
github.com/mxschmitt/playwright-go v0.1100.0
// ...
```

If there's a workaround for this, I'd like to know it.  I did try saving cookies, and loading them into context, but auth says NO and requires login.

Is there a way to regenerate with this option or are there any plans on adding it? 

Thanks!

Also, when trying to regenerate the files, I'm struggling to find the 
`playwright/utils/doclint/generateGoApi.js`

dug through the playwright repo commits with no luck.  How does this work?

Using this dockerfile to create a build env:
```docker
FROM mcr.microsoft.com/playwright:bionic

WORKDIR /tmp
RUN curl -LO https://golang.org/dl/go1.16.3.linux-amd64.tar.gz
RUN rm -rf /usr/local/go && tar -C /usr/local -xzf go1.16.3.linux-amd64.tar.gz
RUN echo 'export PATH=/usr/local/go/bin:$PATH' >> /root/.bashrc
    

WORKDIR /build

COPY  . .
# ms/playwright is cloned in the same dir as go-playwright and moved to / during the build.
RUN mv playwright ..


ENTRYPOINT [ "/bin/bash" ]
```

Running 
```
root@abf47fdd0385:/build# ./scripts/generate-api.sh 
Generating Interfaces
Generated Interfaces
internal/modules/cjs/loader.js:883
  throw err;
  ^

Error: Cannot find module '/playwright/utils/doclint/generateGoApi.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

```
root@abf47fdd0385:/playwright# find . -iname generateGoApi.js
root@abf47fdd0385:/playwright# 
```

--- a/browser.go
+++ b/browser.go
@@ -1,7 +1,9 @@
 package playwright
 
 import (
+	"encoding/json"
 	"fmt"
+	"io/ioutil"
 )
 
 type browserImpl struct {
@@ -20,9 +22,24 @@ func (b *browserImpl) IsConnected() bool {
 
 func (b *browserImpl) NewContext(options ...BrowserNewContextOptions) (BrowserContext, error) {
 	overrides := map[string]interface{}{"sdkLanguage": "javascript"}
-	if len(options) == 1 && options[0].ExtraHttpHeaders != nil {
-		overrides["extraHTTPHeaders"] = serializeHeaders(options[0].ExtraHttpHeaders)
-		options[0].ExtraHttpHeaders = nil
+	if len(options) == 1 {
+		if options[0].ExtraHttpHeaders != nil {
+			overrides["extraHTTPHeaders"] = serializeHeaders(options[0].ExtraHttpHeaders)
+			options[0].ExtraHttpHeaders = nil
+		}
+		if options[0].StorageStatePath != nil {
+			var storageState *BrowserNewContextOptionsStorageState
+			storageString, err := ioutil.ReadFile(*options[0].StorageStatePath)
+			if err != nil {
+				return nil, fmt.Errorf("could not read storage state file: %w", err)
+			}
+			err = json.Unmarshal(storageString, &storageState)
+			if err != nil {
+				return nil, fmt.Errorf("could not parse storage state file: %w", err)
+			}
+			options[0].StorageState = storageState
+			options[0].StorageStatePath = nil
+		}
 	}
 	channel, err := b.channel.Send("newContext", overrides, options)
 	if err != nil {

--- a/browser_context.go
+++ b/browser_context.go
@@ -294,11 +294,14 @@ func (b *browserContextImpl) StorageState(paths ...string) (*StorageState, error
 		return nil, err
 	}
 	if len(paths) == 1 {
-		file, err := os.Open(paths[0])
+		file, err := os.Create(paths[0])
 		if err != nil {
 			return nil, err
 		}
-		if err := json.NewDecoder(file).Decode(result); err != nil {
+		if err := json.NewEncoder(file).Encode(result); err != nil {
+			return nil, err
+		}
+		if err := file.Close(); err != nil {
 			return nil, err
 		}
 	}

--- a/generated-structs.go
+++ b/generated-structs.go
@@ -41,6 +41,10 @@ type BrowserNewContextOptions struct {
 	ReducedMotion *ReducedMotion `json:"reducedMotion"`
 	// Emulates consistent window screen size available inside web page via `window.screen`. Is only used when the [`option: viewport`] is set.
 	Screen *BrowserNewContextOptionsScreen `json:"screen"`
+	// Populates context with given storage state. This option can be used to initialize context with logged-in information obtained via BrowserContext.StorageState(). Either a path to the file with saved storage, or an object with the following fields:
+	StorageState *BrowserNewContextOptionsStorageState `json:"storageState"`
+	// Populates context with given storage state. This option can be used to initialize context with logged-in information obtained via BrowserContext.StorageState(). Path to the file with saved storage state.
+	StorageStatePath *string `json:"storageStatePath"`
 	// Changes the timezone of the context. See [ICU's metaZones.txt](https://cs.chromium.org/chromium/src/third_party/icu/source/data/misc/metaZones.txt?rcl=faee8bc70570192d82d2978a71e2a615788597d1) for a list of supported timezone IDs.
 	TimezoneId *string `json:"timezoneId"`
 	// Specific user agent to use in this context.
@@ -81,6 +85,12 @@ type BrowserScreen struct {
 	Width *int `json:"width"`
 	// page height in pixels.
 	Height *int `json:"height"`
+}
+type BrowserStorageState struct {
+	// Optional cookies to set for context
+	Cookies []BrowserStorageStateCookies `json:"cookies"`
+	// Optional localStorage to set for context
+	Origins []BrowserStorageStateOrigins `json:"origins"`
 }
 type BrowserViewport struct {
 	// page width in pixels.
@@ -129,6 +139,10 @@ type BrowserNewPageOptions struct {
 	ReducedMotion *ReducedMotion `json:"reducedMotion"`
 	// Emulates consistent window screen size available inside web page via `window.screen`. Is only used when the [`option: viewport`] is set.
 	Screen *BrowserNewPageOptionsScreen `json:"screen"`
+	// Populates context with given storage state. This option can be used to initialize context with logged-in information obtained via BrowserContext.StorageState(). Either a path to the file with saved storage, or an object with the following fields:
+	StorageState *BrowserNewPageOptionsStorageState `json:"storageState"`
+	// Populates context with given storage state. This option can be used to initialize context with logged-in information obtained via BrowserContext.StorageState(). Path to the file with saved storage state.
+	StorageStatePath *string `json:"storageStatePath"`
 	// Changes the timezone of the context. See [ICU's metaZones.txt](https://cs.chromium.org/chromium/src/third_party/icu/source/data/misc/metaZones.txt?rcl=faee8bc70570192d82d2978a71e2a615788597d1) for a list of supported timezone IDs.
 	TimezoneId *string `json:"timezoneId"`
 	// Specific user agent to use in this context.
@@ -1467,6 +1481,12 @@ type BrowserNewContextOptionsScreen struct {
 	// page height in pixels.
 	Height *int `json:"height"`
 }
+type BrowserNewContextOptionsStorageState struct {
+	// Optional cookies to set for context
+	Cookies []BrowserNewContextOptionsStorageStateCookies `json:"cookies"`
+	// Optional localStorage to set for context
+	Origins []BrowserNewContextOptionsStorageStateOrigins `json:"origins"`
+}
 type BrowserNewContextOptionsViewport struct {
 	// page width in pixels.
 	Width *int `json:"width"`
@@ -1478,6 +1498,28 @@ type BrowserRecordVideoSize struct {
 	Width *int `json:"width"`
 	// Video frame height.
 	Height *int `json:"height"`
+}
+type BrowserStorageStateCookies struct {
+	Name  *string `json:"name"`
+	Value *string `json:"value"`
+	// Optional either url or domain / path are required
+	URL *string `json:"url"`
+	// Optional either url or domain / path are required
+	Domain *string `json:"domain"`
+	// Optional either url or domain / path are required
+	Path *string `json:"path"`
+	// Optional Unix time in seconds.
+	Expires *float64 `json:"expires"`
+	// Optional httpOnly flag
+	HttpOnly *bool `json:"httpOnly"`
+	// Optional secure flag
+	Secure *bool `json:"secure"`
+	// Optional sameSite flag
+	SameSite *SameSiteAttribute `json:"sameSite"`
+}
+type BrowserStorageStateOrigins struct {
+	Origin       *string                                  `json:"origin"`
+	LocalStorage []BrowserStorageStateOriginsLocalStorage `json:"localStorage"`
 }
 type BrowserNewPageOptionsGeolocation struct {
 	// Latitude between -90 and 90.
@@ -1512,6 +1554,12 @@ type BrowserNewPageOptionsScreen struct {
 	Width *int `json:"width"`
 	// page height in pixels.
 	Height *int `json:"height"`
+}
+type BrowserNewPageOptionsStorageState struct {
+	// Optional cookies to set for context
+	Cookies []BrowserNewPageOptionsStorageStateCookies `json:"cookies"`
+	// Optional localStorage to set for context
+	Origins []BrowserNewPageOptionsStorageStateOrigins `json:"origins"`
 }
 type BrowserNewPageOptionsViewport struct {
 	// page width in pixels.
@@ -1688,11 +1736,59 @@ type BrowserNewContextOptionsRecordVideoSize struct {
 	// Video frame height.
 	Height *int `json:"height"`
 }
+type BrowserNewContextOptionsStorageStateCookies struct {
+	Name  *string `json:"name"`
+	Value *string `json:"value"`
+	// Optional either url or domain / path are required
+	URL *string `json:"url"`
+	// Optional either url or domain / path are required
+	Domain *string `json:"domain"`
+	// Optional either url or domain / path are required
+	Path *string `json:"path"`
+	// Optional Unix time in seconds.
+	Expires *float64 `json:"expires"`
+	// Optional httpOnly flag
+	HttpOnly *bool `json:"httpOnly"`
+	// Optional secure flag
+	Secure *bool `json:"secure"`
+	// Optional sameSite flag
+	SameSite *SameSiteAttribute `json:"sameSite"`
+}
+type BrowserNewContextOptionsStorageStateOrigins struct {
+	Origin       *string                                                   `json:"origin"`
+	LocalStorage []BrowserNewContextOptionsStorageStateOriginsLocalStorage `json:"localStorage"`
+}
+type BrowserStorageStateOriginsLocalStorage struct {
+	Name  *string `json:"name"`
+	Value *string `json:"value"`
+}
 type BrowserNewPageOptionsRecordVideoSize struct {
 	// Video frame width.
 	Width *int `json:"width"`
 	// Video frame height.
 	Height *int `json:"height"`
+}
+type BrowserNewPageOptionsStorageStateCookies struct {
+	Name  *string `json:"name"`
+	Value *string `json:"value"`
+	// Optional either url or domain / path are required
+	URL *string `json:"url"`
+	// Optional either url or domain / path are required
+	Domain *string `json:"domain"`
+	// Optional either url or domain / path are required
+	Path *string `json:"path"`
+	// Optional Unix time in seconds.
+	Expires *float64 `json:"expires"`
+	// Optional httpOnly flag
+	HttpOnly *bool `json:"httpOnly"`
+	// Optional secure flag
+	Secure *bool `json:"secure"`
+	// Optional sameSite flag
+	SameSite *SameSiteAttribute `json:"sameSite"`
+}
+type BrowserNewPageOptionsStorageStateOrigins struct {
+	Origin       *string                                                `json:"origin"`
+	LocalStorage []BrowserNewPageOptionsStorageStateOriginsLocalStorage `json:"localStorage"`
 }
 type BrowserContextStorageStateResultOriginsLocalStorage struct {
 	Name  *string `json:"name"`
@@ -1703,4 +1799,12 @@ type BrowserTypeLaunchPersistentContextOptionsRecordVideoSize struct {
 	Width *int `json:"width"`
 	// Video frame height.
 	Height *int `json:"height"`
+}
+type BrowserNewContextOptionsStorageStateOriginsLocalStorage struct {
+	Name  *string `json:"name"`
+	Value *string `json:"value"`
+}
+type BrowserNewPageOptionsStorageStateOriginsLocalStorage struct {
+	Name  *string `json:"name"`
+	Value *string `json:"value"`
 }

--- a/patches/main.patch
+++ b/patches/main.patch
@@ -192,7 +192,7 @@ index 639324bb2faa3eb9d2c254cbff16f2e380379ab9..96cfbfe8c6ea48c3ae0301cd529713f3
  * langs: csharp, java
  - `body` <[string]>
 diff --git a/playwright/docs/src/api/params.md b/playwright/docs/src/api/params.md
-index f7fd21e5b138fefe026beaeae867212c014257d1..b5aee337efeb7d863f7606296c25d569458d9e74 100644
+index f7fd21e5b138fefe026beaeae867212c014257d1..24f4e9e2485e2e656036ad0e2a9d50ac32204c56 100644
 --- a/playwright/docs/src/api/params.md
 +++ b/playwright/docs/src/api/params.md
 @@ -113,8 +113,8 @@ Defaults to `'visible'`. Can be either:
@@ -206,6 +206,24 @@ index f7fd21e5b138fefe026beaeae867212c014257d1..b5aee337efeb7d863f7606296c25d569
  - `polling` <[float]|"raf">
  
  If [`option: polling`] is `'raf'`, then [`param: expression`] is constantly executed in `requestAnimationFrame`
+@@ -173,7 +173,7 @@ Specify environment variables that will be visible to the browser. Defaults to `
+ Specify environment variables that will be visible to the browser. Defaults to `process.env`.
+ 
+ ## js-python-context-option-storage-state
+-* langs: js, python
++* langs: js, python, go
+ - `storageState` <[path]|[Object]>
+   - `cookies` <[Array]<[Object]>> Optional cookies to set for context
+     - `name` <[string]>
+@@ -202,7 +202,7 @@ Populates context with given storage state. This option can be used to initializ
+ obtained via [`method: BrowserContext.storageState`].
+ 
+ ## csharp-java-context-option-storage-state-path
+-* langs: csharp, java
++* langs: csharp, java, go
+ - `storageStatePath` <[path]>
+ 
+ Populates context with given storage state. This option can be used to initialize context with logged-in information
 @@ -297,7 +297,7 @@ Function to be evaluated in the worker context.
  Function to be evaluated in the worker context.
  

--- a/tests/browser_context_storage_state_test.go
+++ b/tests/browser_context_storage_state_test.go
@@ -1,6 +1,9 @@
 package playwright_test
 
 import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
 	"runtime"
 	"testing"
 
@@ -52,4 +55,100 @@ func TestBrowserContextStorageStateShouldCaptureLocalStorage(t *testing.T) {
 			},
 		},
 	})
+}
+
+func TestBrowserContextStorageStateSetLocalStorage(t *testing.T) {
+	BeforeEach(t)
+	defer AfterEach(t)
+	if isWebKit && runtime.GOOS == "windows" {
+		t.Skip()
+	}
+	context, err := browser.NewContext(
+		playwright.BrowserNewContextOptions{
+			StorageState: &playwright.BrowserNewContextOptionsStorageState{
+				Origins: []playwright.BrowserNewContextOptionsStorageStateOrigins{{
+					Origin: playwright.String("https://www.example.com"),
+					LocalStorage: []playwright.BrowserNewContextOptionsStorageStateOriginsLocalStorage{
+						{
+							Name:  playwright.String("name1"),
+							Value: playwright.String("value1"),
+						},
+					},
+				},
+				},
+			},
+		},
+	)
+	require.NoError(t, err)
+	defer context.Close()
+	page, err := context.NewPage()
+	require.NoError(t, err)
+	defer page.Close()
+	require.NoError(t, page.Route("**/*", func(route playwright.Route, request playwright.Request) {
+
+		require.NoError(t, route.Fulfill(playwright.RouteFulfillOptions{
+			Body: "<html></html>",
+		}))
+	}))
+	_, err = page.Goto("https://www.example.com")
+	require.NoError(t, err)
+	localStorage, err := page.Evaluate("window.localStorage")
+	require.NoError(t, err)
+	require.Equal(t, map[string]interface{}{"name1": "value1"}, localStorage)
+}
+
+func TestBrowserContextStorageStateRoundTripThroughTheFile(t *testing.T) {
+	BeforeEach(t)
+	defer AfterEach(t)
+	page1, err := context.NewPage()
+	require.NoError(t, err)
+	defer page1.Close()
+	require.NoError(t, page1.Route("**/*", func(route playwright.Route, request playwright.Request) {
+		require.NoError(t, route.Fulfill(playwright.RouteFulfillOptions{
+			Body: "<html></html>",
+		}))
+	}))
+	_, err = page1.Goto("https://www.example.com")
+	require.NoError(t, err)
+	_, err = page1.Evaluate(`
+	() => {
+		localStorage["name1"] = "value1"
+		document.cookie = "username=John Doe"
+		return document.cookie
+	}
+	`)
+	require.NoError(t, err)
+	tempfile, err := ioutil.TempFile(os.TempDir(), "storage-state*.json")
+	require.NoError(t, err)
+	state, err := context.StorageState(tempfile.Name())
+	require.NoError(t, err)
+	stateWritten, err := ioutil.ReadFile(tempfile.Name())
+	require.NoError(t, err)
+	var storageState *playwright.StorageState
+	err = json.Unmarshal(stateWritten, &storageState)
+	require.NoError(t, err)
+	require.Equal(t, state, storageState)
+
+	context2, err := browser.NewContext(
+		playwright.BrowserNewContextOptions{
+			StorageStatePath: playwright.String(tempfile.Name()),
+		})
+	require.NoError(t, err)
+	defer context2.Close()
+	page2, err := context2.NewPage()
+	require.NoError(t, err)
+	defer page2.Close()
+	require.NoError(t, page2.Route("**/*", func(route playwright.Route, request playwright.Request) {
+		require.NoError(t, route.Fulfill(playwright.RouteFulfillOptions{
+			Body: "<html></html>",
+		}))
+	}))
+	_, err = page2.Goto("https://www.example.com")
+	require.NoError(t, err)
+	cookie, err := page2.Evaluate("document.cookie")
+	require.NoError(t, err)
+	require.Equal(t, "username=John Doe", cookie)
+	localStorage, err := page2.Evaluate("window.localStorage")
+	require.NoError(t, err)
+	require.Equal(t, map[string]interface{}{"name1": "value1"}, localStorage)
 }

